src/server/worker/api_server.go

```diff
@@ -457,23 +467,35 @@ func (a *APIServer) runUserCode(ctx context.Context, logger *taggedLogger, envir
 		},
 	}
 	cmd.Dir = a.workingDir
-	err := cmd.Run()
-
-	// Return result
-	if err == nil {
-		return nil
+	err := cmd.Start()
+	if err != nil {
+		return err
 	}
-	// (if err is an acceptable return code, don't return err)
-	if exiterr, ok := err.(*exec.ExitError); ok {
-		if status, ok := exiterr.Sys().(syscall.WaitStatus); ok {
-			for _, returnCode := range a.pipelineInfo.Transform.AcceptReturnCode {
-				if int(returnCode) == status.ExitStatus() {
-					return nil
+	cmdErr := make(chan error, 1)
+	go func() {
+		cmdErr <- cmd.Wait()
+	}()
```

Comment:
So as it exists now if you hit that zombie case the zombie processes are left running but we continue as if they've exited which means when we run the next datum we're going to have mutliple processes running at the same time modifying `/pfs/out`. That doesn't sound great. Also in that case we'll leak the above goro, is that correct? If that's true I think we're better of just blocking in that case which will hopefully lead the user to force restart the pod or something like that.

Do we have a sense for how common that case is?

Reply:
So what I recall from testing is that without code like this, cancelling the context (or putting a deadline on the context ... which calls cancel at the right time) doesn't kill the command.  [This is what is described in this issue](https://github.com/golang/go/issues/18874) So if we just called blocking `Run()` the timeout feature wouldn't work. That was the impetus for this restructure. We wait for the execution to finish even if the deadline is exceeded.

What we can do is in the select case where the context is finished, call `cmd.Process.Kill()` which does its best to kill the process, and would make that goro complete. I'll double check this is the case. I thought as written it does kill the process ... but I'll double check.

I can make a job that keeps appending to a randomly named file in `/pfs/out` ... then using `kubectl exec ... ps` or just catting those files ... I should be able to verify that the process is dead.

This won't handle zombies, but that is already the case on master.

Reply:
So I think what you can do here is instead of calling `cmd.Wait()` call `cmd.Process.Wait()` that waits for just the process to return, it doesn't wait for IO to finish. Then, after that returns, check if `ctx.Done()` is closed, if it's not then you call `cmd.Wait()` to make sure the IO has finished, if it is closed then just return, this will potentially lose you some logs that are actively being written... but I think that's an ok compromise to make at this point.

Reply:
This is an example: https://play.golang.org/p/-mPSMvn7NY (that doesn't run... I suspect because go playground won't let you shell out)

Reply:
Cool makes sense. Nifty example.